zap/src/output/mod.rs

```diff
@@ -1,40 +1,87 @@
 use std::fmt::Display;
 
-use crate::{
-	irgen::{gen_ser, Expr, Stmt, Var},
-	parser::{Ty, TyDecl},
-	util::NumTy,
-};
+use crate::{irgen::Stmt, parser::Ty, util::NumTy};
 
 pub mod client;
 pub mod server;
 
-impl Display for TyDecl {
-	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-		let name = &self.name;
-		let ty = &self.ty;
-
-		writeln!(f, "export type {name} = {ty}")?;
+pub trait Output {
+	fn push(&mut self, s: &str);
+	fn indent(&mut self);
+	fn dedent(&mut self);
+	fn push_indent(&mut self);
 
-		writeln!(f, "function types.write_{name}(value: {name})")?;
+	fn push_line(&mut self, s: &str) {
+		self.push_indent();
+		self.push(s);
+		self.push("\n");
+	}
```

Comment:
This is the wrong order. It should go:
- Newline
- Push Indent
- Push String

to make it consistant with other langs iirc

Reply:
This is actually correct order, we start on an empty line, push line content, and then push a newline to the end of the line. This is why files after being formatted end with an empty line, the last line with content is suffixed with a newline.